### ISSUE TITLE
Is it possible to manually define the link function?

### ISSUE THREAD

**Issue by zhongzhixu** [interpretml/interpret#137]
Thanks for the impressive project! I would like to ask Is it possible to manually define the link function?

**Comment by interpret-ml** [comment]
Hi @zhongzhixu , 

The link functions we currently support are Identity (in ExplainableBoostingRegressor) and Logit (in ExplainableBoostingClassifier). Unfortunately, we don't support alternative/custom link functions at the moment. However, this is something we are interested in developing in the future. 

Do you mind sharing some examples of link functions you would like supported? It would be very helpful to us if we understand the use cases a bit better, to help us prioritize appropriately. 

Thanks again for the question/suggestion!

-InterpretML Team

**Comment by PureMath86** [comment]
This would be nice for starters:

https://pygam.readthedocs.io/en/latest/notebooks/tour_of_pygam.html#Link-function:

**Comment by elinrose** [comment]
What is the status here? I am hoping for a log link option.

**Comment by AlistairLR112** [comment]
I would just like to say that some of my colleagues in the insurance industry are interested in the development of this specific ticket - would be very useful for us to model insurance risk

**Comment by paulbkoch** [comment]
Hi @AlistairLR112 -- I would say that you have excellent timing.  Alternative objectives and link functions will be available in our next release.  If your colleagues in insurance are interested in trying it out before then, the functionality is available today in our develop branch.  There's a new parameter for the \_\_init\_\_ functions called "objective" to specify this.  Thanks to a recent contribution by one of our users, the following objectives are available: "poisson_deviance" (poisson with early stopping set to minimize deviance), "gamma_deviance" (gamma with early stopping set to minimize deviance), "rmse_log" (rmse with a log link function), "pseudo_huber", "log_loss" for classification, and of course "rmse".

The wheel for that is available in:

https://dev.azure.com/ms/18804368-0dc8-4666-8c13-330c0ae21c11/_apis/build/builds/450817/artifacts?artifactName=bdist&api-version=7.0&%24format=zip

and it would be installed with:

```
pip install interpret_core-0.3.2-py3-none-any.whl[required,debug,notebook,plotly,lime,sensitivity,shap,ebm,linear,decisiontree,treeinterpreter,dash,skoperules,testing]
```

Your colleagues from insurance might also be interested to know that we recently also added post-processed monotonicity.  There's a new function on the ExplainableBoosting classes called monotonize that does this.

**Comment by AlistairLR112** [comment]
Absolutely fantastic, I will let you know how it works out!

**Comment by MMShZ** [comment]
Hello,
Is it possible to add link functions that can constrain the target values between 0 and 100 (or 1) for percentages? like Sigmoid or Logit! or I just appreciate some guidance on where we can add to the above-mentioned library of functions in the source code. Thanks,

**Comment by paulbkoch** [comment]
Hi @MMShZ -- Writing a custom link function currently requires writing a small amount of C++.  I've tried to make this as easy as possible by creating a fully operational example objective that can be easily modified.  The example objective implements MSE regression, and you can invoke it with:

ebm=ExplainableBoostingRegressor(objective="example")

To use your own link function, fill in the CalcMetric, CalcGradient and CalcGradientHessian functions here:

https://github.com/interpretml/interpret/blob/develop/shared/libebm/compute/objectives/ExampleRegressionObjective.hpp#L81-L104

Then compile it with: build.bat (Windows) or build.sh (Linux/Mac).  Once fitted, the python predict function will return raw score values.  To get predictions, apply the inverse link function to the output.

**Comment by MMShZ** [comment]
Thank you @paulbkoch! With this, actually I could define a custom objective function. I can convert the results back to the actual scale as you mentioned by applying the inverse of the function I used in the objective definition. But I am trying to plot the score functions in the contribution graphs and interaction plots with the correct scale as well (not transformed version). So I am trying to applying the inverse of the function I used on the scores (obtained from ebm_global.data) +ebm.intercept_ => 
doing: inv_link(scores+ebm.intercept_ ) and then plot
however, this does not give reasonable numbers. My assumption by doing this was that implications on transformation in y due to the change in objective function is the same as using a link function g in the formula: g(E(y)) = intercept + f(xi)+f(xi,xj). 
so I wonder what am I missing in the conversion I mentioned on scores (inv_link(scores+ebm.intercept_ ))

**Comment by paulbkoch** [comment]
Hi @MMShZ -- Glad you were able to get this working!  👍

Are you defining both your own objective and link, or just the objective?  For link functions, we currently support logit, identity, and log.  If you have anything else, you're expected to do the inverse link transformation yourself, which should be pretty easy but the inv_link function won't do this.  If you re-used the Example objective, then you should note that the Example objective's default link function is identity, which you can see here with the use of "Link_identity":

https://github.com/interpretml/interpret/blob/143feefc3d953ad8f8fec6221a3a2c22eebe749a/shared/libebm/compute/objectives/ExampleRegressionObjective.hpp#L14

Otherwise, you are correct that we follow the same formula as GAMs: g(E(y)) = intercept + f(xi)+f(xi,xj)

**Comment by MMShZ** [comment]
Hi @paulbkoch, I only defined objective as this could bring my predictions down to the range I wished (0 and 1). The link function is identity. Do you mean I should also define the link function to be the same as objective function? when I apply inverse of the function that I defined in objective (defined myself, not the inv_link of the interpret library) to the predictions (as you mentioned before) I get correct results actually, so it meant to me that loss defined by objective during training would correctly constrain the numbers in the range I wish.  
So, considering that, I get correct predictions when applying inverse of the function that I defined to the predictions -> So, EBM is working well. But just for the sake of visualization, the same thing does not work for the scores when I apply inverse function on the scores+ebm.intercept_ and the numbers are way off from the case that I was just using default RMSE loss. Surprisingly, results are more close to RMSE loss case when I apply inverse function on scores only and then add it to ebm.intercept_, i.e. inverse(scores)+ebm.intercept_, which does not make sense mathematically.

**Comment by paulbkoch** [comment]
Hi @MMShZ -- I'm a little confused about the technique being used here.  Is it possible for you to post the contents of your CalcMetric, CalcGradient, and CalcGradientHessian functions here?

If you are using an identity link function, then you shouldn't need to transform the scores to anything since the identity link does nothing.  The sum of your score values plus the intercept should be the prediction itself.

In order to predict values between 0 and 1, I think the typical thing would be to use a logit or probit link in regression, which allows the score values to have the range -inf to +inf.

**Comment by MMShZ** [comment]
Hi @paulbkoch,
By further investigation, I think I made a mistake as I defined sigmoid instead of logit in the objective and in predictions again I applied sigmoid! (but strangely, the results seem correct as they are the same as MSE case if we ignore the few data points that were predicted to be slightly negative in MSE case, which were correctly predicted to be zero in the case that we used sigmoid). By the way, I will put the contents of the functions here.
```
   GPU_DEVICE inline TFloat CalcMetric(const TFloat& score, const TFloat& target) const noexcept {
      const TFloat one = TFloat(1);
      const TFloat prediction = Exp(score) / (one + Exp(score)); 
      const TFloat error = prediction - target;
      return error * error;
   }

   GPU_DEVICE inline TFloat CalcGradient(const TFloat& score, const TFloat& target) const noexcept {
      const TFloat one = TFloat(1);
      const TFloat prediction = Exp(score) / (one + Exp(score)); 
      const TFloat error = prediction - target;
      const TFloat gradient = Two * error * prediction * (one - prediction);
      return gradient;
   }

   GPU_DEVICE inline GradientHessian<TFloat> CalcGradientHessian(const TFloat& score, const TFloat& target) const noexcept {
      const TFloat one = TFloat(1);
      const TFloat prediction = Exp(score) / (one + Exp(score)); 
      const TFloat error = prediction - target;
      const TFloat gradient = Two * error * prediction * (one - prediction);
      const TFloat hessian = Two * prediction * (one - prediction) * (one - prediction + error);
      return MakeGradientHessian(gradient, hessian);
```

However, my question is that in the beginning I mentioned how to define link function and you suggested modifying the objective and then Once fitted, to get predictions, apply the inverse link function to the outputs. So, if link is identity we should not even **apply inverse link** because we are modifying objective only based on your instruction (and not link). And if we should, then the same thing should be done on scores in interaction heatmaps and contribution graphs as well. So, I am a bit confused. Should we also modify link so that it won't be identity, and how?

**Comment by paulbkoch** [comment]
Hi @MMShZ -- Ok, I think I understand this better now.  First, sorry for any confusion.  The CalcMetric, CalcGradient, and CalcGradientHessian functions all combine the calculation of the inverse link and the calculation of the gradients or metrics.  The parameter "score" is a value in additive-space, so you need to apply the inverse link in order to get the predictions before calculating the rest.  Some other packages separate these calculations into a separate link function function ("function" used twice on purpose), and then separately calculate the gradients from the prediction.  There are some objectives where it's more efficient to have the original score, so I did it this way.

Your first two lines where you calculate the "prediction" variable are the inverse link, which in this case is sigmoid, so your link for this objective is the logit link.

I think this means you need to apply the sigmoid function in order to convert the output of the predict function (which is currently computing scores in the additive-space) into your real predictions.

**Comment by MMShZ** [comment]
Thanks @paulbkoch, 
So based on your elaborate explanation and what I understood, defining sigmoid objective means defining logit link at the same time (as you mentioned: 
> your link for this objective is the logit link.
)
Because I would like to apply logit as a link, the current code would work (as it is for sigmoid) and I am applying sigmoid on the predictions to get the real ones. That actually gives me correct final predictions for each data point and I have no problem with that. So far so good. 
So, the only question that remains for me is that why I cannot get the correct score values (not the final predictions but score for each term) by applying sigmoid on the scores (sigmoid(score))  for ** e.g. drawing custom-built global feature importance plot myself instead of the dashboard**, where I use ebm_global.data to extract score numbers myself? In other words, the numbers for mean absolute score in global feature importance plot should be transformed somehow to be in the scale of the real scores but only applying sigmoid on it does not resolve it.
It seems like applying only inverse link on the scores will not give real scores the same as it will for prediction function.

**Comment by paulbkoch** [comment]
Hi @MMShZ -- It should be reasonable to graph sigmoid(score_from_one_feature + score_from_intercept).  That would be identical to post-process removing all the features from the model except for one.  If it's not working for you, please post the code.

**Comment by MMShZ** [comment]
Hi @paulbkoch, 
Actually that does not work. As I mentioned before, when we try to get correct score (not transformed version by the link function) it is reasonable that the numbers should be near to the case we do not use any link (because by using link in my problem, the predictions for only few data points are changed). But doing the following code will give me numbers way off from that case:
feature importance score for the first term: `sigmoid(ebm.explain_global().data(-1)['overall']['scores'][1]+ebm.intercept_)`

However, when I do the following the numbers are very similar to the case that I did not use link function: `sigmoid(ebm.explain_global().data(-1)['overall']['scores'][1])+ebm.intercept_`. I cannot explain why it is the case.

**Comment by paulbkoch** [comment]
Ah, @MMShZ, I see why this isn't working.  "overall" contains the overall feature/term importance scores.  If you want to do this from the explanation object, it would be something like:

sigmoid(ebm.explain_global().data(-1)['specific'][term_index]['scores']+ebm.intercept_)

I would suggest using the ebm object directly though:

sigmoid(ebm.term_scores_[term_index]+ebm.intercept_)

For graphing, you should know that the first and last bins contain the missing value score and the "unknown" bin score, so depending on what you're doing you probably want to strip those first and last bins out (details in https://interpret.ml/docs/ebm-internals-regression.html)

**Comment by MMShZ** [comment]
Thanks @paulbkoch, that was a useful information!
However, the code I shared was indeed for the purpose of overall feature importance score (sorry if my mention of "feature importance score for the first term" did not imply that). For drawing scores in interaction heatmaps and univariate graphs, I am actually doing as you also mentioned. 
If I do not apply link everything is OK and I can reproduce these interaction heatmaps, but when I apply my link function, as I mentioned, only doing 
sigmoid(ebm.explain_global().data(-1)['specific'][term_index]['scores']+ebm.intercept_) 
will not give correct result, and it is surprising that 
sigmoid(ebm.explain_global().data(-1)['specific'][term_index]['scores'])+ebm.intercept_ 
gives more reasonable numbers! But I will try to check the same thing with ebm.term_scores_ as well.

**Comment by paulbkoch** [comment]
@MMShZ, I don't think that applying the link function to the feature/term importance values provides a useful metric.  The importance values are calculated as the sum total of the absolute value of the scores.  You can see the code here that calculates them:

https://github.com/interpretml/interpret/blob/develop/python/interpret-core/interpret/glassbox/_ebm/_ebm.py#L1913-L1918

If you wanted to construct some kind of new importance metric based on the predictions of a model stripped of all features except for the single feature under consideration, you could do that I guess.  To do that you'd have to calculate the prediction for each of the bins using sigmoid(ebm.term_scores_[term_index]+ebm.intercept_), and then calculate whatever new importance metric you had in mind from those values, presumably weighted by the number of samples in each bin.  The addition of the intercept and application of the inverse link must be done before calculating your metric though instead of after, which appears to be what you are doing currently.

**Comment by paulbkoch** [comment]
Another method that may be conceptually simpler would be to use the remove_terms function (https://interpret.ml/docs/python/api/ExplainableBoostingRegressor.html#interpret.glassbox.ExplainableBoostingRegressor.remove_terms) to remove all terms except for the one you are trying to calculate the feature/term importance value for.  If you have predictions working correctly then this simple change should allow you to calculate the predictions on all your samples using the single feature model, and then you can calculate whatever importance metric you had in mind based on the predictions.

**Comment by MMShZ** [comment]
Hi @paulbkoch,
Sorry for the late reply as I was investigating some aspects of the code. Thanks for the information and I guess that resolved my problem!
